### PR TITLE
generate completions

### PR DESCRIPTION
**Pull request by BoostCookie**
EDIT:
This is done now. It solves https://github.com/JuliaLang/juliaup/issues/151. It works the same way as it does in rustup. https://rust-lang.github.io/rustup/installation/index.html

OLD MESSAGE:
> This is some starting effort that can eventually solve https://github.com/JuliaLang/juliaup/issues/151.
> I don't know how the build files are being packaged for the different distributions and operating systems. I'm asking for feedback in this regard.
> 
> What happens now:
> Every time the package is build the completions for the different shells are generated and put into the build directory `target/release/build/juliaup-<somehexstring>/out/`.
> 
> What is left to do:
> Add the completions to the packages so that they are available after installing juliaup.

### CODE REVIEW

**Comment by davidanthoff** [comment]
I'm not really familiar with how completion files work. Do we need to drop them into a specific location on user systems? Or do they need to be run? Or something else?

**Comment by BoostCookie** [comment]
> I'm not really familiar with how completion files work. Do we need to drop them into a specific location on user systems? Or do they need to be run? Or something else?

Packages usually install their zsh completions to `/usr/share/zsh/site-functions/`. If you are an admin that installs stuff manually you should use `/usr/local/` instead of `/usr/` of course.
Similarly bash completions go into `/usr/share/bash-completion/completions/`.

So OS repositories (Arch User Repository, openSUSE Tumbleweed, Solus) should definitely package the completions into these directories.

For the "recommended" install method
```
curl -fsSL https://install.julialang.org | sh
```
we are only the local user. We could put the completions into `$JULIA_DEPOT_PATH` and then do something similar like here https://github.com/JuliaLang/juliaup/blob/a9d67559c42ddf87598dc80c808674d2381fa623/src/operations.rs#L790 where we add the binary to the `$PATH`. We can also add our completions directory to the specific shell completions.

Though personally I'm not a fan in fishing around in people's config files. I think we should do it the way rustup does it.
https://rust-lang.github.io/rustup/installation/index.html
With a command
```
$ juliaup completions bash > ~/.local/share/bash-completion/completions/juliaup
```
for bash, for example.

**Comment by BoostCookie** [comment]
I have now followed the rustup way, and added a command that generates the completions.

File: src/bin/juliaup.rs
Comment (fingolfin) [inline src/bin/juliaup.rs:30]: Disclaim: I am not a maintainer of this repository, so feel free to ignore me. This advice is just based on my experience maintaining other projects. If this PR was one on my repository, I'd be reluctant to review and merge it as it is somewhat opaque.

Specifically: Do you *need* to move these definitions to a new file? If not, I would recommend against doing it, at least in this PR, as it makes it that much harder to review it, as it is not obvious if anything (and what) was changed in this code block other than moving it.


If the code must be split out, I would recommend doing it in two commits: the first commit just moves the code to a new file, but without otherwise modifying it. The second commit than does the actual work and modifications.

Comment (BoostCookie) [reply src/bin/juliaup.rs:30]: Thanks for your feedback.
Unfortunately I think separating the files is necessary, otherwise the `command_completions.rs` cannot access the `enum Juliaup`.

I've now separated the commits like you recommended.

**Comment by fingolfin** [comment]
@BoostCookie unfortunately it seems this now as conflicts that must be resolved?

**Comment by BoostCookie** [comment]
Is there even any interest in this pull request?
@davidanthoff

**Comment by StefanKarpinski** [comment]
Yes, this is great functionality. @davidanthoff is quite busy, but please don't be discouraged. We'll get this merged.

**Comment by davidanthoff** [comment]
We had more changes to the clap structures on `main`, and I'm a bit worried that those changes might get lost in the simultaneous move of the code to a new file in this PR. So, I just created a new PR off `main` that _just_ moves the clap structs to a new file, like in this PR. I'm going to merge that right away, and then we can be entirely sure that we don't lose anything in the merge story.

@BoostCookie could you rebase this PR on top of `main` once I've merged https://github.com/JuliaLang/juliaup/pull/787?

Awesome PR, thanks!

**Review by davidanthoff — APPROVED** [review]

**Pull request merged**
